File: src/ast/ArrayType.java
package ast;

import java.io.StringWriter;

public class ArrayType extends BaseNode {
    private BaseNode _base;
    private BaseNode _dimensionExpression;
    private String _dimensionString;

    public ArrayType(BaseNode Base) {
        super(NodeType.ArrayType);
        _base = Base;
        _dimensionExpression = null;
    }

    public ArrayType(BaseNode Base, BaseNode dimensionExpression) {
        super(NodeType.ArrayType);
        _base = Base;
        _dimensionExpression = dimensionExpression;
    }

    public ArrayType(BaseNode Base, String dimensionString) {
        super(NodeType.ArrayType);
        _base = Base;
        _dimensionString = dimensionString;
    }

    @Override
    public boolean HasRightPart() {
        return true;
    }

    @Override
    public boolean IsArray() {
        return true;
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        _base.PrintLeft(writer);
    }

    @Override
    public void PrintRight(StringWriter writer) {
        // FIXME: detect if previous char was a ].
        writer.write(" ");

        writer.write("[");

        if (_dimensionString != null) {
            writer.write(_dimensionString);
        } else if (_dimensionExpression != null) {
            _dimensionExpression.Print(writer);
        }

        writer.write("]");

        _base.PrintRight(writer);
    }
}

File: src/de/fearlesstobi/demangler/ast/ArrayType.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class ArrayType : BaseNode {
    private val base: BaseNode
    private var dimensionExpression: BaseNode? = null
    private var dimensionString: String? = null

    constructor(base: BaseNode) : super(NodeType.ArrayType) {
        this.base = base
        dimensionExpression = null
    }

    constructor(base: BaseNode, dimensionExpression: BaseNode?) : super(NodeType.ArrayType) {
        this.base = base
        this.dimensionExpression = dimensionExpression
    }

    constructor(base: BaseNode, dimensionString: String?) : super(NodeType.ArrayType) {
        this.base = base
        this.dimensionString = dimensionString
    }

    public override fun hasRightPart(): Boolean {
        return true
    }

    override val isArray: Boolean
        get() = true

    public override fun printLeft(writer: StringWriter) {
        base.printLeft(writer)
    }

    public override fun printRight(writer: StringWriter) { // FIXME: detect if previous char was a ].
        writer.write(" ")
        writer.write("[")
        if (dimensionString != null) {
            writer.write(dimensionString)
        } else if (dimensionExpression != null) {
            dimensionExpression!!.print(writer)
        }
        writer.write("]")
        base.printRight(writer)
    }
}